lib/Sema/TypeCheckStmt.cpp

```diff
@@ -2519,6 +2526,11 @@ bool TypeCheckASTNodeAtLocRequest::evaluate(
       if (isa<TapExpr>(E))
         return Action::SkipChildren(E);
 
+      // Don't walk into SingleValueStmtExprs, they should be type-checked as
+      // a whole.
+      if (isa<SingleValueStmtExpr>(E))
+        return Action::SkipChildren(E);
+
```

Comment:
Just for understanding: This is the main change, right?

But if the `if` expression contains multiple statements, I think we do want to walk into it, right? E.g when you have one of the following.

```swift
enum Foo: Error {
  case bar
}

func test(_ x: Int) {}

let a: Foo = if true {
  test(#^COMPLETE^#)
  throw Foo.bar
} else {
  Foo.bar
}
```

Or even worse

```swift
enum Foo: Error {
  case bar
}

func test(_ x: Int) {}

let a: Foo = if true {
  func localFunc() {
    test(#^COMPLETE^#)
  }
  throw Foo.bar
} else {
  Foo.bar
}
```

In those cases all we want to type-check is `test(#^COMPLETE^#)` and not the entire `if` expression.

Reply:
Yeah this is the main change, we could definitely make it more efficient yeah, we really only need to type-check as a whole if the completion is in a single-expression branch

Reply:
> and even then we could still potentially skip type-checking non-expression branches

I think we should be skipping unrelated conjunctions when type checking for completion. I worked on this for result builders and it could also apply to if expressions, I just never checked. 

---

Could you add tests for the cases above. I agree that it’s a very uncommon thing to define a function inside an if expression, so if we aren’t taking the most optimal path, that’s probably OK. but we should definitely make sure that we do produce results.

Reply:
> I think we should be skipping unrelated conjunctions when type checking for completion

Yeah, I saw that, though I think we can be even more aggressive in this case and prune entire branches. Updated to include that, and added the extra test cases.